Add textFitWidth and textFitHeight to sprites

DavidBuerer opened the pull request:
This PR is to re-open https://github.com/maplibre/maplibre-gl-js/pull/1800 with the changes recommended by @zekefarwell and picked by vote.  To quote the original PR: This is to provide a way to work around shields with small text getting too small (left) without creating multiple layers with multiple copies of the shields.

![image](https://github.com/maplibre/maplibre-gl-js/assets/29717748/dd4d23d0-6d5b-4c26-b808-4bd88a08b58c)

To do this, a textFitWidth and textFitHeight property are added to sprites.  The anticipated use (besides setting both to stretchOrShrink which is the same as not adding them) is to set textFitWidth=stretchOnly and textFitHeight=proportional

The following is the table of results using narrow content and wide content for the property.
<img width="250" alt="image" src="https://github.com/maplibre/maplibre-gl-js/assets/29717748/1b301aec-2abc-46be-9d9e-2ba05c9bc13a">
Refreshing my memory on the proposal I came to a different understanding of the implementation, so this table is different than the one I had come up with manually before.  Note that if neither axis is `proportional`, there is basically no effect.  Let me know if I am interpreting this incorrectly and if `stretchOnly` makes sense without `proportional`.  Note also that setting both to `proportional` requires a default axis.  I have it set to height, which leads to some odd results in longer strings as per the table.

HarelM commented:
Thanks for taking the time to open this PR!
Can you please open an issue in maplibre-style-spec repo with the description here and the documentation updates of the sprite json? 
This will allow us to both have all the spec changes in one place and also allow native library to follow up with the implementation. 

Also please check why the tests fail.

DavidBuerer commented:
> Thanks for taking the time to open this PR! Can you please open an issue in maplibre-style-spec repo with the description here and the documentation updates of the sprite json? This will allow us to both have all the spec changes in one place and also allow native library to follow up with the implementation.
> 
> Also please check why the tests fail.

Will do.  

Not sure why the ubuntu rendering is different.  Will have to get a VM spun up on Monday to look at it.  Is there a way to see the results.html that the server build generated?

Also not clear on why the package is that much bigger.  Any thoughts there?

HarelM commented:
The build summary has the html file with the images generated from the linux worker.
The images are base64 encoded, so you can extract them and commit them.
But I'm not sure I understand why existing tests would fail...

HarelM commented on src/style/style_image.ts:
```diff
@@ -31,6 +31,27 @@ export type StyleImageData = {
+/**
+ * Enumeration of possible values for StyleImageMetadata.textFitWidth and textFitHeight.
+ */
+export enum TextFit {
```
Should be const enum, to avoid enum quirkiness.

HarelM commented on src/symbol/shaping.ts:
```diff
@@ -802,6 +803,49 @@ function shapeIcon(
+ * @param shapedIcon - The icon that will be adjusted.
+ * @returns Extents of the shapedIcon with text fit adjustments if necessary.
+ */
+function applyTextFit(shapedIcon) {
```
Please add unit tests for this new method.

HarelM commented on test/examples/textFit.html:
```diff
@@ -0,0 +1,150 @@
+<html lang="en">
+
+<head>
+    <title>MapLibre GL JS debug page</title>
```
This will be added to the examples, if this is intended, please add a better title and og:description, otherwise delete this page.

DavidBuerer replied:
Removed

HarelM commented on test/integration/render/tests/icon-text-fit/textFit-anchors-long/README.md:
```diff
@@ -0,0 +1,2 @@
+This an adaptation of stretch-nine-part-content but with a long
```
I believe you can use the metadata in the style.json file for this kind of comments.

HarelM commented:
I would consider keeping the existing tests as is and create new tests. This would make sure this is a non breaking change. 
If this is a breaking change, then it can only be part of a breaking change version which might happen at September the earliest, as we have released a breaking change version at the beginning of this year...

DavidBuerer commented:
> I would consider keeping the existing tests as is and create new tests. This would make sure this is a non breaking change. If this is a breaking change, then it can only be part of a breaking change version which might happen at September the earliest, as we have released a breaking change version at the beginning of this year...

That was my bad.  Sometimes unit tests are alongside their targets and sometimes in tests.  Didn't realize this was the later.  Moved tests over.

HarelM commented on src/render/image_atlas.ts:
```diff
@@ -4,6 +4,7 @@ import {register} from '../util/web_worker_transfer';
 import potpack from 'potpack';
 
 import type {StyleImage} from '../style/style_image';
+import {TextFit} from '../style/style_image';
```
`impoer type ...`?

DavidBuerer replied:
It looks like enums cannot be treated as types.

HarelM commented on test/integration/symbol-shaping/shaping.test.ts:
```diff
@@ -389,3 +389,77 @@ describe('fitIconToText', () => {
 
 });
+
+describe('applyTextFit', () => {
```
I think you can move this test to be next to `shaping.ts` i.e. `shaping.test.ts` as you are unit testing the method.
Some other tests in this file (not sure if all of them) are using the file system to compare the results of the test and in some cases to update the test results.
But the tests here are unit tests and doesn't use the file system at all as far as I understand.

HarelM replied:
Feel free to move other tests that are not using any json files to that file.

DavidBuerer replied:
You already asked me to move it out, but I can move it back.

HarelM replied:
I don't think I did, but the important part from my point of view is to keep integration tests in integration folder and unit tests next to the file, it's sometimes not a clear cut obviously, and I might have said something that got lost in translation.
Thanks for taking the time to do so!

HarelM commented on test/integration/render/tests/icon-text-fit/textFit-grid-long/style.json:
```diff
@@ -0,0 +1,182 @@
+      "test": {
+        "width": 360,
+        "height": 180,
+        "description": "This draws the matrix of all possible combinations of textFitWidth (columns) and textFitHeight (rows) where they proceed from the top-left in the order stretchOrShrink, stretchOnly, and proportional.  This version shows the effect of a long string with a wider aspect ratio than the source image."
```
Can you add a render test where the text is from top to bottom to see the effect when the text stretches the icon downwards?

DavidBuerer replied:
Added textFit-grid-long-vertical

DavidBuerer replied:
Also added textFit-grid-short-vertical.  The results are somewhat unexpected to me... but could still be "correct".

HarelM commented on src/symbol/shaping.ts:
```diff
@@ -802,6 +803,56 @@ function shapeIcon(
+ * @param shapedIcon - The icon that will be adjusted.
+ * @returns Extents of the shapedIcon with text fit adjustments if necessary.
+ */
+function applyTextFit(shapedIcon: PositionedIcon): Box {
```
I'd like to see the coverage on this method once the unit test are moved, as I don't have coverage for the integration tests, only on render and unit tests.

DavidBuerer replied:
They have been moved.  Is there action you want me to take?

HarelM replied:
Improve coverage as stated below.

HarelM commented on src/symbol/quads.ts:
```diff
@@ -65,8 +65,12 @@ export function getIconQuads(
 
-    const iconWidth = shapedIcon.right - shapedIcon.left;
-    const iconHeight = shapedIcon.bottom - shapedIcon.top;
+    let icon = {
```
```suggestion
    let icon: Box = {
```

DavidBuerer replied:
Done

HarelM commented on src/symbol/shaping.ts:
```diff
@@ -802,6 +803,56 @@ function shapeIcon(
     return {image, top: y1, bottom: y2, left: x1, right: x2};
 }
 
+export interface Box {
```
Isn't `PositionedIcon` good enough? Why do we need this extra type?

DavidBuerer replied:
applyText fit returns only the box (x1,y1,x2,y2).  If it returned a PositionedIcon, then it returns an image and collisionPadding also.  If space was a concern, I could see it, but it is a type that goes away once transpiled.

HarelM commented on src/symbol/quads.ts:
```diff
@@ -88,28 +92,39 @@ export function getIconQuads(
 
     if (image.content && hasIconTextFit) {
         const content = image.content;
+        const contentWidth = content[2] - content[0];
```
Is this where the stretching happens? Maybe we can have the two methods of stretch and fit in the same file/place?

DavidBuerer replied:
It could.  I attempted to lay it out so that if you don't set textFitWidth or textFitHeight the original logic was preserved; to reduce the chance that this change affects mainstream scenarios.

HarelM replied:
That's the right approach, no argue here, I was thinking more about refactoring of moving similar logic to the same file.

DavidBuerer commented:
The images for textFit-grid-long generate differently on my machine than on the render server.  I can "fix" this by using the images from the server, but then they may not run locally successfully.  I am continuing to investigate why this is.
![image](https://github.com/maplibre/maplibre-gl-js/assets/29717748/d5902f9d-84ac-4be7-9900-696960d129e8)

HarelM commented:
There are differences between machines, what we usually do is add more than one image to a folder so that it passes locally and on the CI server.

HarelM commented:
Based on the coverage report (although the result may change when all the render test pass) there are still some lines missing. 
Can you add a test to cover the missing lines (I'm not talking about partial if yellow lines, but rather missing red lines in the codecov website or the jest report).

HarelM commented:
A few last comments: 
1. Please update the build size to pass the build test
2. Please see if there are more places that needs tests, as I see there are some coverage gaps in newly added lines
3. Please open a PR in the maplibre-style-spec repo to update the documentation about sprites, please make sure to add a sdk support table to the new parameters alongside some of the explanations here. I just finished refactoring these docs and it should be fairly easy to update them (look for sprite in the docs folder there).

Thanks!!

DavidBuerer commented:
> A few last comments:
> 
> 1. Please update the build size to pass the build test
> 2. Please see if there are more places that needs tests, as I see there are some coverage gaps in newly added lines
> 3. Please open a PR in the maplibre-style-spec repo to update the documentation about sprites, please make sure to add a sdk support table to the new parameters alongside some of the explanations here. I just finished refactoring these docs and it should be fairly easy to update them (look for sprite in the docs folder there).
> 
> Thanks!!

Updated build size and addressed missing coverage lines.  Will start in on the docs.

DavidBuerer commented:
Is there anything else that I have to do, or am I waiting for someone with write access to complete?

DavidBuerer commented:
> The style spec repo changes I mentioned above.

Working on that, but that has to be separate PR.  You want that in place before completing this?

HarelM commented:
Yes, to make sure that when this is merged and released the style spec is updated as well. 
Sorry for linking the two, but I think it's for the best.

DavidBuerer commented:
https://github.com/maplibre/maplibre-style-spec/pull/637
